lib/src/generate/build_impl.dart

```diff
@@ -85,15 +84,21 @@ class BuildImpl {
   /// Deletes all previous output files.
   Future _deletePreviousOutputs() async {
     if (await _reader.hasInput(_buildOutputsId)) {
-      // Cache file exists, just delete all outputs contained in it.
+      /// Cache file exists, delete all outputs which don't appear in the
+      /// [_assetGraph], or are marked as needing an update.
+      ///
+      /// Removes all files from [_inputsByPackage] regardless of state.
       var previousOutputs =
           JSON.decode(await _reader.readAsString(_buildOutputsId));
       await _writer.delete(_buildOutputsId);
       _inputsByPackage[_buildOutputsId.package]?.remove(_buildOutputsId);
-      await Future.wait(previousOutputs.map((output) {
+      await Future.wait(previousOutputs.map((output) async {
         var outputId = new AssetId.deserialize(output);
         _inputsByPackage[outputId.package]?.remove(outputId);
-        return _writer.delete(outputId);
+        var node = _assetGraph.get(outputId);
+        if (node == null || (node as GeneratedAssetNode).needsUpdate) {
```

Comment:
Will this always be a safe cast to `GeneratedAssetNode`? If so, why not type `_assetGraph`?